src/envs/cover.py

```diff
@@ -345,3 +345,372 @@ def __init__(self) -> None:
+    controls the gripper "magnet" or "vacuum".
+
+    The state space also needs to change to track x, y, grip.
+
+    The notion of allowed hand regions needs to be adjusted.
+    We will allow the gripper to move anywhere, except when
+    the y dimension is less than a certain threshold, then
+    the x dimension will still need to be within a hand region.
+
+    Currently simulate disallows dropping blocks except when
+    over a target region. We should fix this to allow dropping
+    blocks anywhere. Also, would be good to check collisions.
+    """
+    hand_region_height = 0.1
+    grasp_height_tol = 1e-2
+    grasp_thresh = 0.
+    initial_block_y = 0.05
+    block_height = 0.1
+    target_height = 0.1
+    initial_robot_y = 0.4
+
+    def __init__(self) -> None:
+        super().__init__()
+        # Need to now include y and gripper info in state.
+        # Removing "pose" because that's ambiguous.
+        # Also adding height to blocks.
+        # The y position will correspond to the top of the block.
+        self._block_type = Type(
+            "block", ["is_block", "is_target", "width", "x",
+                      "grasp", "y", "height"])
+        # Targets don't need y because they're constant.
+        self._target_type = Type(
+            "target", ["is_block", "is_target", "width", "x"])
+        # Also removing "hand" because that's ambiguous.
+        self._robot_type = Type("robot", ["x", "y", "grip"])
+        # Need to override predicate creation because the types are
+        # now different (in terms of equality).
+        self._IsBlock = Predicate(
+            "IsBlock", [self._block_type], self._IsBlock_holds)
+        self._IsTarget = Predicate(
+            "IsTarget", [self._target_type], self._IsTarget_holds)
+        self._Covers = Predicate(
+            "Covers", [self._block_type, self._target_type], self._Covers_holds)
+        self._HandEmpty = Predicate(
+            "HandEmpty", [], self._HandEmpty_holds)
+        self._Holding = Predicate(
+            "Holding", [self._block_type], self._Holding_holds)
+        # Need to override object creation because the types are now
+        # different (in terms of equality).
+        self._blocks = []
+        self._targets = []
+        for i in range(CFG.cover_num_blocks):
+            self._blocks.append(Object(f"block{i}", self._block_type))
+        for i in range(CFG.cover_num_targets):
+            self._targets.append(Object(f"target{i}", self._target_type))
+        self._robot = Object("robby", self._robot_type)
+        # Override the original options to make them multi-step
+        self._Pick = ParameterizedOption(
+            "Pick", types=[self._block_type],
+            params_space=Box(-0.1, 0.1, (1,)),
+            _policy=self._Pick_policy,
+            _initiable=self._Pick_initiable,
+            _terminal=self._Pick_terminal)
+        # Note: there is a change here -- the parameter space is now
```

Comment:
Why did you make this change? I'm personally curious, but also you should clarify that in this comment.

Reply:
We made this change because we want to learn options whose parameters correspond to changes in the low-level states of the objects involved in the option's effects. Added to the comment